### DISCUSSION TITLE
Smooth scaling not matching equation in literature

### DISCUSSION THREAD

**Discussion by jqloh** [dfki-ric/movement_primitives#53] [General]
![image](https://github.com/dfki-ric/movement_primitives/assets/169635940/727bbf3a-ba8e-464a-86ea-ac12a5a3afaf)
the K (or alpha_y*beta_y) term is missing from the code as specified by equation 6 in Learning and Generalization of Motor Skills by Learning from Demonstration, is there a reason for this?

same goes for the determine_force code snippet as well. 
![image](https://github.com/dfki-ric/movement_primitives/assets/169635940/5cabb3ba-c80b-4447-9f32-e427d70de1e7)
should it not be divided by K as well

**Comment by AlexanderFabisch** [comment]
> the K (or alpha_y*beta_y) term is missing from the code as specified by equation 6 in Learning and Generalization of Motor Skills by Learning from Demonstration, is there a reason for this?

Where exactly is it missing? For the forcing term? That doesn't make a big difference. It is just a constant scaling factor for the weights of the forcing term. If you are interested, you can play around with it and look at this example script: https://github.com/dfki-ric/movement_primitives/blob/main/examples/plot_dmp_scaling.py

(Relevant PR: https://github.com/dfki-ric/movement_primitives/pull/33)

**Comment by jqloh** [comment]
Hi Alex, thanks for the reply

Yes, I meant the forcing term. According to Learning and Generalization of Motor Skills by Learning from Demonstration Equation 6, the equation for acceleration and the forcing terms (during imitation) is as such
![image](https://github.com/dfki-ric/movement_primitives/assets/169635940/6d9f7134-043f-4ad7-bea2-041e6ff75a29)

However, in the _dmp.py file, I believe **_dmp_acc function** and the **determine forces function** seems to have missed considering the K constant from the equation. 

I've tried to follow the paper and did some modifications to the _dmp.py file as shown below to include the K term from the forces

determine_forces
![image](https://github.com/dfki-ric/movement_primitives/assets/169635940/629b9338-d5f9-4c4c-897a-9650f378a63a)
_dmp_acc 
![image](https://github.com/dfki-ric/movement_primitives/assets/169635940/ddf88d79-701a-4829-9446-be78bfb7e8bb)


However, when I ran the modification, I got this plot
![image](https://github.com/dfki-ric/movement_primitives/assets/169635940/237cfaa5-74fe-4c43-a505-e172841aa46e)

was wondering if you could take a look at it

**Comment by AlexanderFabisch** [comment]
If you got a matching pair of step function and "determine forces" function, it should work. Are you sure that indentation is correct?

But as I said before, it does not make a difference whether you scale by K or not. This will only change the scale of the weights in the end. The behavior should be exactly the same.

**Comment by jqloh** [comment]
I've doubled checked the step function and determine forces function several times, however, it still gave me the exact results. Are there any locations that I'm missing other than the _dmp_acc function used by the step function and the "determine forces" function? 

I plan to employ this library for my PhD hence why I'm trying to map what I'm reading from this godsent library with the literature.

**Comment by AlexanderFabisch** [comment]
The default DMP has four implementations of the step function: Python, Cython, Euler integration, RK4 integration and combinations of these. RK4 integration in Cython is used by default. If you changed the Python function, make sure that it is also used, e.g. with

```python
_, Y_dmp = dmp.open_loop(step_function="rk4")
```

If you changed the Cython version, you also have to recompile it.

But you should definitely check which function is used.

**Comment by jqloh** [comment]
Ahh, it was indeed an issue with the Cython package, I did not change the dmp_fast.pyx, hence it was using the _dmp_acc from dmp_fast.pyx and the determine_forces from python, thus the weird results. It was working as required when I changed both Cython and Python. 

Side note, when I reverted the smoothing_scaling == False case to the original equation (by including the (goal-start)*force ) set by Ijspeert, once I resolved the Cython and Python issue.

Thank you so much, and have a great day!
